checker/src/main/java/org/checkerframework/checker/guieffect/GuiEffectVisitor.java

```diff
@@ -222,20 +225,33 @@ private void lambdaAssignmentCheck(
     @Override
     public Void visitVariable(VariableTree node, Void p) {
         Void result = super.visitVariable(node, p);
-        if (node.getInitializer() != null
-                && node.getInitializer().getKind() == Tree.Kind.LAMBDA_EXPRESSION) {
-            lambdaAssignmentCheck(
-                    atypeFactory.getAnnotatedType(node),
-                    (LambdaExpressionTree) node.getInitializer(),
-                    "assignment.type.incompatible");
+        if (node.getInitializer() != null) {
+            if (node.getInitializer().getKind() == Tree.Kind.NEW_CLASS) {
+                commonAssignmentCheck(
```

Comment:
Why is this special logic needed? Isn't super already checking that an initializer has a suitable type for the variable?
Why only if the top-level expression is NEW_CLASS and not if e.g. there is a ternary expression that contains a NEW_CLASS?

Reply:
@wmdietl There are two parts to this:

1) The reason this special logic is needed is because inner class and lambda effect inference happen during the visit. Basically, I only know if a lambda/class should be inferred to be `@UI` after I have visited its methods and seen if they call in turn a method which has `@UIEffect` (either directly or based on e.g. polymorphism). Because of the order of the traversal in `BaseTypeVisitor`, I believe the lambda/class won't have been marked `@UI` by the time the variable assignment is evaluated inside the call to `super.visitVariable(...)`.

2) You are absolutely right, this breaks down for ternary expressions, and I suspect other expressions which can contain a new class as well. I am currently trying to figure this out, since even just unconditionally calling `commonAssignmentCheck` for all nodes is not enough for a ternary expression that includes one instance with inferred `@UI` type and one instance that is `@AlwaysSafe`